crates/bevy_ecs/src/schedule_v3/schedule.rs

```diff
@@ -0,0 +1,1022 @@
+    /// This prevents overflow and thus prevents false positives.
+    ///
+    /// **Note:** Does nothing if the [`World`] counter has not been incremented at least [`CHECK_TICK_THRESHOLD`](crate::change_detection::CHECK_TICK_THRESHOLD)
+    /// times since the previous pass.
+    pub(crate) fn check_change_ticks(&mut self, change_tick: u32) {
+        #[cfg(feature = "trace")]
+        let _span = bevy_utils::tracing::info_span!("check stored schedule ticks").entered();
+        // label used when trace feature is enabled
+        #[allow(unused_variables)]
+        for (label, schedule) in self.inner.iter_mut() {
+            #[cfg(feature = "trace")]
+            let name = format!("{:?}", label);
+            #[cfg(feature = "trace")]
+            let _span =
+                bevy_utils::tracing::info_span!("check schedule ticks", name = &name).entered();
+            schedule.check_change_ticks(change_tick);
+        }
+    }
+}
+
+/// A collection of systems, and the metadata and executor needed to run them
+/// in a certain order under certain conditions.
+pub struct Schedule {
+    graph: ScheduleMeta,
+    executable: SystemSchedule,
+    executor: Box<dyn SystemExecutor>,
+}
+
+impl Default for Schedule {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl Schedule {
+    /// Constructs an empty `Schedule`.
+    pub fn new() -> Self {
+        Self {
+            graph: ScheduleMeta::new(),
+            executable: SystemSchedule::new(),
+            executor: Box::new(MultiThreadedExecutor::new()),
+        }
+    }
+
+    /// Add a system to the schedule.
+    pub fn add_system<P>(&mut self, system: impl IntoSystemConfig<P>) -> &mut Self {
+        self.graph.add_system(system);
+        self
+    }
+
+    /// Add a collection of systems to the schedule.
+    pub fn add_systems<P>(&mut self, systems: impl IntoSystemConfigs<P>) -> &mut Self {
+        self.graph.add_systems(systems);
+        self
+    }
+
+    /// Configure a system set in this schedule.
+    pub fn configure_set(&mut self, set: impl IntoSystemSetConfig) -> &mut Self {
+        self.graph.configure_set(set);
+        self
+    }
+
+    /// Configure a collection of system sets in this schedule.
+    pub fn configure_sets(&mut self, sets: impl IntoSystemSetConfigs) -> &mut Self {
+        self.graph.configure_sets(sets);
+        self
+    }
+
+    /// Sets the system set that new systems and system sets will join by default
+    /// if they aren't already part of one.
+    pub fn set_default_set(&mut self, set: impl SystemSet) -> &mut Self {
+        self.graph.set_default_set(set);
+        self
+    }
+
+    /// Returns the schedule's current execution strategy.
+    pub fn get_executor_kind(&self) -> ExecutorKind {
+        self.executor.kind()
+    }
+
+    /// Sets the schedule's execution strategy.
+    pub fn set_executor_kind(&mut self, executor: ExecutorKind) -> &mut Self {
```

Comment:
This would likely require us to make `SystemSchedule` pub, which would require the `unsafe impl Sync` on it to hold under any executor.

Reply:
Yeah, neither `SystemExecutor` nor `SystemSchedule` are `pub`. To present a nice trait for users to implement themselves, I think all of this would need some extensive changes. (Not enthusiastic about doing that.)

The whole pipeline (app builder -> configs -> schedule -> executable -> executor) is pretty coupled here. Seems like we'd have to `pub` everything involved for users to really make something custom, not just the very last step.

Reply:
Reasonable. I guess for now I'm cool with using this abstraction and if the need arises we can revisit. The enum does make the setting options centralized + easily discoverable.